src/linux/build_id_reader.rs

```diff
@@ -0,0 +1,225 @@
+            &program_headers_data,
+            0,
+            self.header.e_phnum as usize,
+            self.context,
+        )?;
+        for header in program_headers {
+            if header.p_type != elf::program_header::PT_NOTE {
+                continue;
+            }
+            if let Ok(Some(result)) =
+                self.find_build_id_note(header.p_offset, header.p_filesz, header.p_align)
+            {
+                return Ok(result);
+            }
+        }
+        Err(Error::NoProgramHeaderNote)
+    }
+
+    /// Read the build id from a notes section.
+    pub fn read_from_section(&self) -> Result<Vec<u8>, Error> {
+        if self.header.e_shoff == 0 {
+            return Err(Error::NoSectionNote);
+        }
+
+        let section_headers_data = read(
+            &self.module_memory,
+            self.header.e_shoff,
+            self.header.e_shentsize as u64 * self.header.e_shnum as u64,
+        )?;
+        let section_headers = elf::SectionHeader::parse(
+            &section_headers_data,
+            0,
+            self.header.e_shnum as usize,
+            self.context,
+        )?;
+        let strtab_section_header = &section_headers[self.header.e_shstrndx as usize];
+
+        for header in &section_headers {
+            let sh_name = header.sh_name as u64;
+            if sh_name >= strtab_section_header.sh_size {
+                log::warn!("invalid sh_name offset");
+                continue;
+            }
+            if sh_name + NOTE_SECTION_NAME.len() as u64 >= strtab_section_header.sh_size {
+                // This can't be a match.
+                continue;
+            }
+            let name = read(
+                &self.module_memory,
+                strtab_section_header.sh_offset + sh_name,
+                NOTE_SECTION_NAME.len() as u64,
+            )?;
+            if NOTE_SECTION_NAME == &*name {
+                return match self.find_build_id_note(
+                    header.sh_offset,
+                    header.sh_size,
+                    header.sh_addralign,
+                ) {
+                    Ok(Some(v)) => Ok(v),
+                    Ok(None) => Err(Error::NoSectionNote),
+                    Err(e) => Err(e),
+                };
+            }
+        }
+
+        Err(Error::NoSectionNote)
+    }
+
+    /// Generate a build id by hashing the first page of the text section.
+    pub fn generate_from_text(&self) -> Result<Vec<u8>, Error> {
+        if self.header.e_shoff == 0 {
+            return Err(Error::NoTextSection);
+        }
+
+        let section_headers_data = read(
+            &self.module_memory,
+            self.header.e_shoff,
+            self.header.e_shentsize as u64 * self.header.e_shnum as u64,
+        )?;
+        let section_headers = elf::SectionHeader::parse(
+            &section_headers_data,
+            0,
+            self.header.e_shnum as usize,
+            self.context,
+        )?;
```

Comment:
I actually almost did that and cached the `section_headers`. I'll probably do that; I didn't because I thought the likelihood of not having a build id was so low it didn't matter much anyway (and I wanted to keep these methods `&self` rather than `&mut self`).